tensornetwork/network_components.py

```diff
@@ -232,6 +237,82 @@ def __str__(self) -> Text:
     return self.name
 
 
+class CopyNode(Node):
+  def __init__(self,
+               rank: int,
+               dimension: int,
+               name: Text,
+               axis_names: List[Text],
+               backend: base_backend.BaseBackend) -> None:
+    # TODO: Make this computation lazy, once Node doesn't require tensor
+    # at instatiation.
+    copy_tensor = self.make_copy_tensor(rank, dimension)
+    copy_tensor = backend.convert_to_tensor(copy_tensor)
+    super().__init__(copy_tensor, name, axis_names, backend)
+
+  @staticmethod
+  def make_copy_tensor(rank: int, dimension: int) -> Tensor:
+    shape = (dimension,) * rank
+    copy_tensor = np.zeros(shape)
+    i = np.arange(dimension)
+    copy_tensor[(i,) * rank] = 1
+    return copy_tensor
+
+  def _is_my_trace(self, edge: "Edge") -> bool:
+    return edge.node1 is self and edge.node2 is self
+
+  def _get_partner(self, edge: "Edge") -> Tuple[Node, int]:
+    if edge.node1 is self:
+        return edge.node2, edge.axis2
+    assert edge.node2 is self
+    return edge.node1, edge.axis1
+
+  def _get_partners(self) -> List[Tuple[Node, int]]:
+    partners = []
+    for edge in self.edges:
+      if edge.is_dangling() or self._is_my_trace(edge):
+        continue
+      partner_node, shared_axis = self._get_partner(edge)
+      partners.append((partner_node, shared_axis))
+    return partners
+
+  _VALID_LETTERS = list('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ')
+
+  def _make_einsum_input_term(self,
+                              node: Node,
+                              shared_axis: int,
+                              next_index: int) -> Tuple[str, int]:
+    indices = []
+    for axis in range(node.get_rank()):
+      if axis == shared_axis:
+        indices.append(0)
+      else:
+        indices.append(next_index)
+        next_index += 1
+    term = "".join(self._VALID_LETTERS[i] for i in indices)
+    return term, next_index
+
+  def _make_einsum_output_term(self, next_index: int) -> str:
+    return "".join(self._VALID_LETTERS[i] for i in range(1, next_index))
+
+  def _make_einsum_expression(self, partners: List[Tuple[Node, int]]) -> str:
+    next_index = 1  # zero is reserved for the shared index
+    einsum_input_terms = []
+    for partner_node, shared_axis in partners:
+      einsum_input_term, next_index = self._make_einsum_input_term(
+              partner_node, shared_axis, next_index)
+      einsum_input_terms.append(einsum_input_term)
+    einsum_output_term = self._make_einsum_output_term(next_index)
+    einsum_expression = ",".join(einsum_input_terms) + "->" + einsum_output_term
+    return einsum_expression
+
+  def contract_all_edges(self) -> Tensor:
```

Comment:
This really shouldn't be part of `CopyNode`. We've kept the interface to be `net.contract_*` and we did that to allow all of the nodes and edges to be updated automatically. Since this only calculates the contracted tensor, maybe name it something like `calculated_full_contraction` and add `contract_copy_tensor` to TensorNetwork.

Reply:
I agree. This method doesn't make any of the re-arrangements of the network that contraction entails. Instead, it computes the tensor that we'll need in the implementation of `contract_copy_tensor()` in `TensorNetwork`. I plan this as a follow-up PR.

Renamed.

Reply:
Aww I see. Ok cool sgtm. I would still recommend a name change but the structure seems ok.

Reply:
I did rename it :-)

New name: `compute_contracted_tensor()`.

Reply:
Yes you did. My bad.